### PR TITLE
Replace Repo with RepoRegex implementation

### PR DESCRIPTION
**Pull request by cursedcoder**
@keegancsmith here's a rough idea how to implement the query syntax for repo regex.

I know you added something similar with RepoRegex last month – haven't touched that.

Let me know what you think.

### CODE REVIEW

**Review by keegancsmith — COMMENTED** [review]
The thing I was most interested in was your ideas around changes to the syntax. So it seems this change is changing `repo:expr` where `expr` was always substring to `expr` is now always a regex. I think this is a valid change. We should document this breaking change though.

Otherwise when it comes to actual implementation, I think it would be clearer if it works more like `query.RepoRegexp` where instead of having a child `query.Q`, the thing inside of `Repo` is always a regexp object we can use. Also get the benefit of not having to compile it per matchtree.

cc @sourcegraph/search-core

**Comment by cursedcoder** [comment]
@keegancsmith I investigated the code further in the regard of the compiled regexp optimization you suggested.

The way google folks (or whoever was the author) designed the code suggests we probably have to use `query.Regex` instead.

There is [Regex.FileName](https://github.com/sourcegraph/zoekt/blob/master/query/query.go#L89) boolean which impacts several places in the codebase:

* [Query parser creates filename regex](https://github.com/sourcegraph/zoekt/blob/ae8862730b7d6847471f6f39d86f1f95bdcbec67/query/parse.go#L146) - see third parameter
* [Regex matchtree takes the fileName param](https://github.com/sourcegraph/zoekt/blob/ae8862730b7d6847471f6f39d86f1f95bdcbec67/eval.go#L534)
* [Regex matchtree generates match candidates with the boolean](https://github.com/sourcegraph/zoekt/blob/ae8862730b7d6847471f6f39d86f1f95bdcbec67/matchtree.go#L660)
* [Finally making a line match in content provider](https://github.com/sourcegraph/zoekt/blob/master/contentprovider.go#L137-L140)

To keep the same semantics we would also want to produce the fragment matches for the repo. 
What bothers me is ambiguous usage of repo name as file name here.

Does it make sense?

**Comment by keegancsmith** [comment]
We don't need matchtree support for repository queries. We evaluate repository queries before that time (via `indexData.simplify`). So we shouldn't be guided by exactly how `query.Regexp` and `query.Substring` work.

This branch makes it so the query language now always expects a regular expression for `repo:` queries. I think we should reflect that in the implementation of `query.Repo` such that it contains a regex.

Note: the reason I like using `regexp.Regexp` is because that makes it easy to call a find/match function on it vs `syntax.Regexp`. `query.Regexp` uses `syntax.Regexp` because the matchtree logic needs the AST to generate substring queries. We don't need this for repository evaluation and can just rely on regexp.

I think I may be confusing here / not providing enough info. I'm happy to explore this idea further and push to this branch something a little more concrete to demonstrate what I am thinking. Tell me what ya think. Also sorry about the slow reply. I've adjusted my inbox filtering rules to better account for outside contributors :)

**Comment by cursedcoder** [comment]
@keegancsmith ok now it looks exactly like RepoRegex with query parsing added.

Let me know how to correctly handle BC and what to do about RepoRegex (if it's needed).

File: query/parse.go [hunk 1]
```diff
@@ -124,7 +125,11 @@ func parseExpr(in []byte) (Q, int, error) {
 		}
 		expr = &caseQ{text}
 	case tokRepo:
-		expr = &Repo{Pattern: text}
+		if text == "" {
+			return nil, 0, fmt.Errorf("the repo: expr must have an argument")
+		}
```
Comment (keegancsmith) [inline query/parse.go]: `repo:` is a valid query, it matches everything. I think the empty string compiles as a regex? If not, maybe special case it to `.*`?

File: query/parse.go [hunk 2]
```diff
@@ -124,7 +125,11 @@ func parseExpr(in []byte) (Q, int, error) {
...
+
+		expr = &Repo{regexp.MustCompile(text)}
```
Comment (keegancsmith) [inline query/parse.go]: you can handle the error here and return it rather than using MustCompile which can panic.

File: query/query_test.go [hunk 3]
```diff
@@ -84,7 +84,7 @@ func TestSimplify(t *testing.T) {
 }
 
 func TestMap(t *testing.T) {
-	in := NewAnd(&Substring{Pattern: "bla"}, &Not{&Repo{"foo"}})
+	in := NewAnd(&Substring{Pattern: "bla"}, &Not{&Repo{&Substring{Pattern: "foo"}}})
```
Comment (keegancsmith) [inline query/query_test.go]: this test looks out of date?

File: web/server.go [hunk 4]
```diff
@@ -564,7 +564,7 @@ func (s *Server) servePrintErr(w http.ResponseWriter, r *http.Request) error {
 	}
 	qs := []query.Q{
 		&query.Regexp{Regexp: re, FileName: true, CaseSensitive: true},
-		&query.Repo{Pattern: repoStr},
+		&query.Repo{Expr: &query.Substring{Pattern: repoStr}},
```
Comment (keegancsmith) [inline web/server.go]: this test looks out of date?

**Review by keegancsmith — COMMENTED** [review]
This LGTM to me. Some inline feedback and I think this is ready to be marked as ready to review.

Don't worry about RepoRegexp, I'll follow up and remove it. We are currently using it internally, I can move Sourcegraph over to Repo and then remove RepoRegexp.

**Comment by cursedcoder** [comment]
@keegancsmith should be better now

File: web/server.go [hunk 5]
```diff
@@ -562,9 +562,16 @@ func (s *Server) servePrintErr(w http.ResponseWriter, r *http.Request) error {
 	if err != nil {
 		return err
 	}
+
+	repoRe, err := regexp.Compile(repoStr)
```
Comment (keegancsmith) [inline web/server.go]:
```suggestion
	repoRe, err := regexp.Compile("^"+regexp.QuoteMeta(repoStr)+"$")
```

**Review by keegancsmith — APPROVED** [review]
LGTM. One small bug inline to fix. Also I think you may need to rebase since master has moved a bit since this PR was created.

**Review by keegancsmith — APPROVED** [review]
LGTM!

**Comment by keegancsmith** [comment]
@cursedcoder Can you run `go test ./...` and fix up the compilation errors? Sorry I see we are missing CI on this PR.

**Comment by cursedcoder** [comment]
Will do once I'm home. But I ran the tests before pushing it last time.

**Comment by keegancsmith** [comment]
I'm getting compilation failures. Easy to fix though, just can't push to this branch :)

```patch
diff --git a/eval.go b/eval.go
index 0eacd6f..a325deb 100644
--- a/eval.go
+++ b/eval.go
@@ -18,6 +18,7 @@ import (
        "context"
        "fmt"
        "log"
+       "regexp"
        "regexp/syntax"
        "sort"
        "strings"

```

**Comment by cursedcoder** [comment]
Ok will fix in 45 mins!

**Comment by cursedcoder** [comment]
@keegancsmith:

```
⋊> ~/s/g/s/zoekt on query-reporegex-parse  go test ./...                                                                            16:02:07
ok  	github.com/google/zoekt	4.237s
ok  	github.com/google/zoekt/build	2.327s
?   	github.com/google/zoekt/cmd	[no test files]
?   	github.com/google/zoekt/cmd/zoekt	[no test files]
ok  	github.com/google/zoekt/cmd/zoekt-archive-index	2.281s
?   	github.com/google/zoekt/cmd/zoekt-git-clone	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-git-index	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-hg-index	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-index	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-indexserver	[no test files]
ok  	github.com/google/zoekt/cmd/zoekt-merge-index	1.076s
?   	github.com/google/zoekt/cmd/zoekt-mirror-bitbucket-server	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-mirror-gerrit	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-mirror-github	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-mirror-gitiles	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-mirror-gitlab	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-repo-index	[no test files]
ok  	github.com/google/zoekt/cmd/zoekt-sourcegraph-indexserver	3.899s
?   	github.com/google/zoekt/cmd/zoekt-test	[no test files]
?   	github.com/google/zoekt/cmd/zoekt-webserver	[no test files]
ok  	github.com/google/zoekt/ctags	(cached)
?   	github.com/google/zoekt/debugserver	[no test files]
ok  	github.com/google/zoekt/gitindex	18.444s
ok  	github.com/google/zoekt/ignore	(cached)
?   	github.com/google/zoekt/internal/mockSearcher	[no test files]
ok  	github.com/google/zoekt/query	(cached)
ok  	github.com/google/zoekt/rpc	0.658s
?   	github.com/google/zoekt/rpc/internal/srv	[no test files]
ok  	github.com/google/zoekt/shards	2.701s
ok  	github.com/google/zoekt/stream	0.918s
?   	github.com/google/zoekt/trace	[no test files]
ok  	github.com/google/zoekt/web	1.300s
```

**Pull request merged**